### PR TITLE
CASMINST-4256 update ntp module with better airgap support

### PR DESCRIPTION
**Pull request by jacobsalmela**
Signed-off-by: Jacob Salmela <jacob.salmela@hpe.com>

#### Summary and Scope

- Partially automates CASMINST-4256
- Partially implements MTL-1635
 
##### Issue Type

- RFE Pull Request

Adds additional support for airgap environments, which doubles as the fix for existing bugs on older systems.  Specifically, it automates step 11 here: https://github.com/Cray-HPE/docs-csm/pull/1139/files

#### Prerequisites

- [x] I have included documentation in my PR (or it is not required)
- [x] I tested this on internal system (x) (if yes, please include results or a description of the test)
 

#### Idempotency
 
Ran the code over and over and checked the files contents.

```
ncn-w003:~ # ls -l /etc/chrony.d/
total 4
-rw-r--r-- 1 root root 726 Mar 30 10:36 cray.conf
-rw-r--r-- 1 root root   0 Mar 30 10:37 cray.conf.dist
-rw-r--r-- 1 root root   0 Mar 30 10:37 pool.conf
-rw-r--r-- 1 root root   0 Mar 30 10:37 removeme.dist
ncn-w003:~ # grep iburst /etc/chrony.conf && cloud-init single --name ntp --frequency always && grep iburst /etc/chrony.conf
# ! pool pool.ntp.org iburst
Cloud-init v. 20.2-8.45.1 running 'single' at Wed, 30 Mar 2022 10:37:59 +0000. Up 83749.48 seconds.
# ! pool pool.ntp.org iburst
ncn-w003:~ # ls -l /etc/chrony.d/
total 4
-rw-r--r-- 1 root root 726 Mar 30 10:38 cray.conf
ncn-w003:~ #
ncn-w003:~ # grep iburst /etc/chrony.conf && cloud-init single --name ntp --frequency always && grep iburst /etc/chrony.conf
# ! pool pool.ntp.org iburst
Cloud-init v. 20.2-8.45.1 running 'single' at Wed, 30 Mar 2022 10:39:51 +0000. Up 83860.56 seconds.
# ! pool pool.ntp.org iburst
ncn-w003:~ # ls -l /etc/chrony.d/
total 4
-rw-r--r-- 1 root root 726 Mar 30 10:39 cray.conf
ncn-w003:~ #
```

checked the logs:

```
...
...
2022-03-30 10:38:00,075 - util.py[DEBUG]: Writing to /etc/chrony.conf - wb: [640] 1187 bytes
2022-03-30 10:38:00,076 - cc_ntp.py[DEBUG]: Airgap: Problematic config found: /etc/chrony.d/pool.conf
2022-03-30 10:38:00,076 - util.py[DEBUG]: Attempting to remove /etc/chrony.d/pool.conf
2022-03-30 10:38:00,076 - cc_ntp.py[DEBUG]: Airgap: Problematic config found: /etc/chrony.d/removeme.dist
2022-03-30 10:38:00,076 - util.py[DEBUG]: Attempting to remove /etc/chrony.d/removeme.dist
2022-03-30 10:38:00,076 - cc_ntp.py[DEBUG]: Airgap: Problematic config found: /etc/chrony.d/cray.conf.dist
2022-03-30 10:38:00,076 - util.py[DEBUG]: Attempting to remove /etc/chrony.d/cray.conf.dist
...
...
2022-03-30 10:41:23,053 - util.py[DEBUG]: Read 1187 bytes from /etc/chrony.conf
2022-03-30 10:41:23,053 - cc_ntp.py[DEBUG]: Airgap: Commenting out pool directive from /etc/chrony.conf
2022-03-30 10:41:23,053 - util.py[DEBUG]: Writing to /etc/chrony.conf - wb: [640] 1187 bytes
2022-03-30 10:41:23,054 - cc_ntp.py[DEBUG]: Airgap: Problematic config found: /etc/chrony.d/cray.conf.dist
2022-03-30 10:41:23,054 - util.py[DEBUG]: Attempting to remove /etc/chrony.d/cray.conf.dist
...
...
```
 
#### Risks and Mitigations
 
Low.  Tested on 1.0.x and 1.2.x systems.

### CODE REVIEW

File: cloudinit/config/cc_ntp.py [hunk 1]
```diff
@@ -464,6 +466,80 @@ def rename_ntp_conf(confpath=None):
...
+                    'Airgap: Problematic config found: %s', c
+                )
+                # remove each .dist file
+                util.del_file(c)
+
+
+def comment_default_pool(confpath=None):
+    """Comments out the default pool in the chrony configuration file.
+
+    @param confdir: string. Specify a path to a .conf file
```
Comment (manderson-hpe) [inline cloudinit/config/cc_ntp.py:494]: Should this be `confpath`?

File: cloudinit/config/cc_ntp.py [hunk 2]
```diff
@@ -464,6 +466,80 @@ def rename_ntp_conf(confpath=None):
...
+            nopool,
+            mode=0o640,
+        )
+
+
+def remove_pool_confd(confpath=None):
+    """Delete the pool.conf config, which has unreachable servers in
+    airgap environments.
+
+    @param confdir: string. Specify a path to a .conf file
```
Comment (manderson-hpe) [inline cloudinit/config/cc_ntp.py:522]: confpath

File: cloudinit/config/cc_ntp.py [hunk 3]
```diff
@@ -464,6 +466,80 @@ def rename_ntp_conf(confpath=None):
...
+    """
+    if os.path.exists(confpath):
+        LOG.debug(
+            'Airgap: Problematic config found: %s', confpath
```
Comment (manderson-hpe) [inline cloudinit/config/cc_ntp.py:526]: Should probably say `Problematic config file removed:`

**Review by manderson-hpe — CHANGES_REQUESTED** [review]

File: cloudinit/config/cc_ntp.py [hunk 4]
```diff
@@ -47,6 +47,8 @@
 
 frequency = PER_INSTANCE
 NTP_CONF = "/etc/ntp.conf"
+CHRONY_CONF = "/etc/chrony.conf"
```
Comment (rustydb) [inline cloudinit/config/cc_ntp.py:50]: Should these be alphabetized?
Like this
```python
CHRONY_CONF = "/etc/chrony.conf"
CHRONY_POOL_CONF = "/etc/chrony.d/pool.conf"
NTP_CONF = "/etc/ntp.conf"
NR_POOL_SERVERS = 4
```

File: cloudinit/config/cc_ntp.py [hunk 5]
```diff
@@ -464,6 +466,80 @@ def rename_ntp_conf(confpath=None):
...
+        # find the *.dist files in the given folder
+        dist_files = util.glob.glob(os.path.join(confdir, "*.dist"))
+        for c in dist_files:
+            if dist_files is None:
+                LOG.debug(
+                    'Airgap: No *.dist files found to remove'
+                )
+            else:
+                LOG.debug(
+                    'Airgap: Problematic config found: %s', c
```
[the remark this answers is not in the document]
Comment (rustydb) [reply cloudinit/config/cc_ntp.py:485]: I agree, since the next line runs `util.del_file(c)`.

File: cloudinit/config/cc_ntp.py (hunk 3 above)
Comment (rustydb) [reply cloudinit/config/cc_ntp.py:526]: It should say it, since the next line actually removes it.

**Review by rustydb — CHANGES_REQUESTED** [review]
I requested a few aesthetic changes, otherwise it looks good.

File: cloudinit/config/cc_ntp.py (hunk 4 above)
Comment (jacobsalmela) [reply cloudinit/config/cc_ntp.py:49]: Fixed

File: cloudinit/config/cc_ntp.py (hunk 1 above)
Comment (jacobsalmela) [reply cloudinit/config/cc_ntp.py]: Fixed

File: cloudinit/config/cc_ntp.py (hunk 2 above)
Comment (jacobsalmela) [reply cloudinit/config/cc_ntp.py]: fixed

File: cloudinit/config/cc_ntp.py (hunk 3 above)
Comment (jacobsalmela) [reply cloudinit/config/cc_ntp.py:526]: > `cc_ntp` alerts us (Cray-HPE) about airgap issues, but doesn't actually take an action. I did this intentionally so it's easier to see what cloud-init is _about_ to do as we often are trying to determine where cloud-init fails.
> 
> ```
> 2022-03-30 10:41:23,054 - cc_ntp.py[DEBUG]: Airgap: Problematic config found: /etc/chrony.d/cray.conf.dist
> ```
> 
> The next call is to `util.del_file`, which is built into cloud-init, which has more info about actually removing the file itself.
> 
> ```
> 2022-03-30 10:41:23,054 - util.py[DEBUG]: Attempting to remove /etc/chrony.d/cray.conf.dist
> ```

**Review by rustydb — APPROVED** [review]

**Pull request merged**